src/geometry/manifold/cont_mdiff.lean

```diff
@@ -2121,3 +2121,87 @@ hf.smul hg
 lemma smooth.smul {f : M → 𝕜} {g : M → V} (hf : smooth I 𝓘(𝕜) f) (hg : smooth I 𝓘(𝕜, V) g) :
   smooth I 𝓘(𝕜, V) (λ p, f p • g p) :=
 hf.smul hg
+
+/-! ### Smoothness of an open embedding and its inverse -/
+
+variable (I)
+lemma cont_mdiff_open_embedding
+  {M : Type*} [topological_space M] [nonempty M]
+  {e : M → H} (h : open_embedding e) {n : with_top ℕ} :
+  @cont_mdiff _ _ _ _ _ _ _ I _ _ h.singleton_charted_space _ _ _ _ _ I _ _ _ n e :=
+begin
+  haveI := h.singleton_smooth_manifold_with_corners I,
+  rw cont_mdiff_iff,
+  split,
+  { apply h.continuous },
+  intros x y, -- show the function is actually the identity on the range of I ∘ e
+  apply cont_diff_on.congr cont_diff_on_id,
+  intros z hz, -- factorise into the chart (=e) and the model (=id)
+  rw [ext_chart_at_coe, ext_chart_at_coe_symm, chart_at_self_eq],
+  repeat {rw function.comp_app},
+  rw [local_homeomorph.refl_apply, id.def, local_homeomorph.singleton_charted_space_chart_at_eq,
+    open_embedding.to_local_homeomorph_right_inv e],
+  { rw model_with_corners.right_inv,
+    { refl },
+    apply set.mem_of_subset_of_mem _ hz.1,
+    apply ext_chart_at_target_subset_range },
```

Comment:
Generally, I think it is better to replace a terminal `apply` by `exact` and possible underscores.